src/ocean/core/Option.d

```diff
@@ -0,0 +1,236 @@
+        ```
+
+        Parameters:
+            cb_undefined = action to take if value is not defined
+            cb_defined   = ditto for defined. May modify internal value via
+                reference
+
+    **************************************************************************/
+
+    public void visit ( void delegate() cb_undefined,
+        void delegate(ref T) cb_defined )
+    {
+        if (this.defined)
+            cb_defined(this.value);
+        else
+            cb_undefined();
+    }
+
+    /**************************************************************************
+
+        A more "old-school" version of `visit`. Provided both to conform style
+        of existing code and avoid delegate bugs in dmd1.
+
+        Discouraged by default as more error-prone than `visit`.
+
+        Parameters:
+            value = out parameter, will be written to with content of `this` if
+            defined
+
+        Return:
+            `true` if this is defined and `value` was updated.
+
+    **************************************************************************/
+
+    public bool get ( ref T value )
+    {
+        if (this.defined) value = this.value;
+        return this.defined;
+    }
+
+    /**************************************************************************
+
+        Returns:
+            `true` if this is defined.
+
+    **************************************************************************/
+
+    public bool isDefined ( )
+    {
+        return this.defined;
+    }
+}
+
+///
+unittest
+{
+    alias Option!(bool) Maybe;
+
+    Maybe x = true;
+    Maybe y = false;
+    Maybe z = Maybe.undefined;
+
+    x.visit(
+        ()               { test(false); },
+        (ref bool value) { test(value); }
+    );
+
+    y.visit(
+        ()               { test(false); },
+        (ref bool value) { test(!value); }
+    );
+
+    z.visit(
+        ()               { test(true); },
+        (ref bool value) { test(false); }
+    );
+}
+
+/******************************************************************************
+
+    Shortcut on top of Option to created defined value, uses IFTI to reduce
+    the noise
+
+    Parameters:
+        value = value to wrap into Option
+
+    Returns:
+        wrapped value
+
+******************************************************************************/
+
+public Option!(T) option ( T ) ( T value )
+{
+    return Option!(T)(value, true);
+}
+
+///
+unittest
+{
+    Option!(int) foo ( bool x )
+    {
+        if (x)
+            return option(42);
+        else
+            return undefined!(int);
+    }
+
+    foo(true).visit(
+        ()              { test(false); },
+        (ref int value) { test(value == 42); }
+    );
+
+    foo(false).visit(
+        ()              { },
+        (ref int value) { test(false); }
+    );
+}
+
+/******************************************************************************
+
+    Shortcut on top of Option to created undefined value.
```

Comment:
`to create` and I feel you're missing some articles here, but I can't place them myself :-)

Reply:
"to create an"